### PR TITLE
wcwidth calculation

### PR DESCRIPTION
**Pull request by jerch**
Quick replacement of `isWide()` with `wcwidth`.  Since xterm.js has no frontend packaging atm I just copied my version into the source file.

wcwidth fixes several issues:
- full unicode (all planes supported at data stage, output depends on font capability)
- combining characters
- half/fullwidth distinction for all unicode codepoints (though lookup table needs fixes to get xterm or font compatibility)

Misaligned glyphs are not fixed by this pull request. This would need a way to introspect the current font and glyph widths. This could be done by prerendering faulty glyphs onto a canvas, getting their width with  `measureText` and aligning them with a CSS rule. I played with this approach here https://github.com/netzkolchose/jquery.browserterminal/blob/master/dist/js/jquery.browserterminal.js#L289, the glyph width calculation is pretty slow but is only needed once.

### CODE REVIEW

**Comment by parisk** [comment]
Hi @jerch, thanks for your contribution! Before moving on with testing this out, could you please mention the particular GitHub issues that this PR is supposed to fix?

**Comment by jerch** [comment]
Issues affected by this: #72 , maybe #70 , #62

File: src/xterm.js [hunk 1]
```diff
@@ -70,6 +70,122 @@
     var window = this, document = this.document;
 
     /**
+     * wcwidth
+     */
+    var wcswidth = (function() {
+        var _WIDTH_COMBINING = [
+            [0x0300, 0x036F], [0x0483, 0x0486], [0x0488, 0x0489],
```
Comment (Tyriar) [inline src/xterm.js]: Where was this list sourced from?

File: src/xterm.js [hunk 2]
```diff
@@ -1468,8 +1589,30 @@
         this.maxRange();
       }
 
+      // apply leftover surrogate left from last write
+      if (this.surrogate_left) {
+            data = this.surrogate_left + data;
```
Comment (Tyriar) [inline src/xterm.js]: Use 2 space indents

File: src/xterm.js [hunk 3]
```diff
@@ -1468,8 +1589,30 @@
...
+        }
+
       for (; i < l; i++) {
         ch = data[i];
+
+        code = data.charCodeAt(i);
+        // surrogate high
+        if (0xD800 <= code && code <= 0xDBFF) {
+            low = data.charCodeAt(i+1);
+            if (low !== low) {  // end of data stream, we need to save to first part of surrogate
```
Comment (Tyriar) [inline src/xterm.js]: `low` will always equal `low`?

File: src/xterm.js [hunk 4]
```diff
@@ -70,6 +70,122 @@
     var window = this, document = this.document;
 
     /**
+     * wcwidth
+     */
+    var wcswidth = (function() {
```
Comment (Tyriar) [inline src/xterm.js]: General helper functions should go at the end of the file.

File: src/xterm.js (hunk 1 above)
Comment (jerch) [inline src/xterm.js]: https://www.cl.cam.ac.uk/%7Emgk25/ucs/wcwidth.c

File: src/xterm.js (hunk 3 above)
Comment (jerch) [inline src/xterm.js]: `charCodeAt` with an illegal index returns NaN, (NaN !== NaN) evaluates always to true.

Comment (Tyriar) [inline src/xterm.js]: Can you use `isNaN()` for clarity?

**Comment by jerch** [comment]
I have cleaned up the code and fixed the indentation.

**Comment by Tyriar** [comment]
Thanks @jerch! I'll have to take a chunk of time to look at this in detail, can you think of some test cases that could be added for `wcwidth`?

**Comment by jerch** [comment]
yes I have some test cases, gonna try to apply them to xterm.js.

There are some edge cases that need proper testing:
- fullwidth chars just go to next line if entered at `cols-2`, what to do with the last cell itself? It will need at least the attributes set.
- combining chars should go to last active cell - needs tests for all cells, no clue yet how solve this after cursor jumps (xterm just insert them at the cursor after jumps)

**Comment by jerch** [comment]
One test case fails - I commented it out since the it refers to a missing feature in default state (wrap true/false is not handled there at all).

**Comment by parisk** [comment]
Okay, just found the chance to take a deeper look into this. #72 does not seem to be closed by this (still weird characters appear, but different than the ones mentioned in this issue).

This PR closes #70 and closes #62 successfully though.

I will move on with merging this PR, since it great steps forward have been done here. I also believe that #72 can be tackled autonomously, since it appears only when moving the cursor.

Thanks a lot!
